assets/js/leaves.js

```diff
@@ -0,0 +1,566 @@
+// logic to generate a path for drawing leaves,
+// and for managing leaf details and tag behavior in the timetree
+
+import { select, selectAll } from "d3-selection";
+import { line, curveNatural } from "d3-shape";
+
+// combine into d3 object for convenience
+const d3 = {
+  line,
+  curveNatural,
+  select,
+  selectAll,
+};
+
+// configuration for leaf sizes
+// sizes are scaled from 40px width design:
+//   min height and width: 40px; max height 100px;
+//   max mid shift 34 (34px at most, 17 either way); max tip shift 12
+
+let leafSize = {
+  minHeight: 30,
+  maxHeight: 75,
+  width: 30,
+  maxMidShift: 25,
+  maxTipShift: 9,
+};
+
+// get a random number between a min and max
+function randomNumBetween(max, min = 0) {
+  // scale random value between 0 and 1 to desired scale, start at min value
+  return Math.random() * (max - min) + min;
+}
+
+function plusOrMinus(x) {
+  return randomNumBetween(x, -x);
+}
+
+function cointoss() {
+  // randomly return true or false;
+  // based on https://stackoverflow.com/a/60322877/9706217
+  return Math.random() < 0.5;
+}
+
+const TagSelectEvent = new Event("tag-select");
+const TagDeselectEvent = new Event("tag-deselect");
+
+class Leaf {
+  // constant for selection classname
+  static selectedClass = "select";
+  static highlightClass = "highlight";
+
+  constructor(panel, ignore_ids, tags) {
+    // store a reference to the panel object
+    this.panel = panel;
+    this.container = document.querySelector("aside");
+    this.bindHandlers();
+    // list of ids to ignore when loading leaf details
+    // (known non-leaf elements, including branch start targets)
+    this.ignore_ids = ignore_ids || [];
+    this.tags = tags || [];
+  }
+
+  static isTag(element) {
+    // check if an element is a tag
+    return (
+      element.tagName == "A" && element.parentElement.classList.contains("tags")
+    );
+  }
+
+  bindHandlers() {
```

Comment:
generally, it's good practice to remove event listeners when they are no longer needed to avoid memory leaks and a progressive slow down of the application. A lot of the `addEventListener`s here are attached to the container (the panel, I believe) so they're not too risky: they are attached when the page loads and are needed active until the user navigates away or refreshes the page.
I wonder if others could be a bit more leaky, however.